src/name_core.jl

```diff
@@ -71,6 +71,30 @@ Base.@pure function dim_noerror(dimnames::Tuple{Vararg{Symbol, N}}, name::Symbol
     return 0
 end
 
+"""
+    expand_dimnames(dimnames, name)
+
+For `dimnames` being a tuple of names (symbols) for the dimensions.
+and `name` being a name.
+This expands the `dimnames` if `name` is not in `dimnames`.
+e.g. `expand_dimnames((:a, :b), :c) == (:a, :b, :c)`
+If `name` is already in `dimnames` then `dimnames` is returned.
+"""
+function expand_dimnames(dimnames::Tuple, name::Symbol)
+    if name in dimnames
+        return dimnames
+    else
+        return (dimnames..., name)
```

Comment:
I don't understand the comment, i.e. what a code constant is and what does it mean for it to fold (hold?). Could you please explain or provide a reference? Sorry for rookie Q.

Reply:
I'll let wikipedia explain
https://en.wikipedia.org/wiki/Constant_folding

Basically we want to know what code is generated for constant.
Which can be be checked using the following pattern:
```julia
julia> f(names) = (names..., :zzz)
f (generic function with 1 method)

julia> @code_typed (()->f((:x,:y)))()
CodeInfo(
1 ─ %1 = Core.tuple(:x, :y, :zzz)::Tuple{Symbol,Symbol,Symbol}
└──      return %1
) => Tuple{Symbol,Symbol,Symbol}
```
see that a literal has ended up in the output.

in contrast:
```julia
julia> g(names) = (names..., Symbol("zzz" * ""))
g (generic function with 1 method)

julia> @code_typed (()->g((:x,:y)))()
CodeInfo(
1 ─ %1 = invoke Base.string("zzz"::String, ""::String)::String
│   %2 = $(Expr(:foreigncall, :(:jl_string_ptr), Ptr{UInt8}, svec(Any), 0, :(:ccall), :(%1)))::Ptr{UInt8}
│   %3 = Core.sizeof(%1)::Int64
│   %4 = $(Expr(:foreigncall, :(:jl_symbol_n), Ref{Symbol}, svec(Ptr{UInt8}, Int64), 0, :(:ccall), :(%2), :(%3), :(%3), :(%2)))::Symbol
│   %5 = Core.tuple(:x, :y, %4)::Tuple{Symbol,Symbol,Symbol}
└──      return %5
) => Tuple{Symbol,Symbol,Symbol}
```

Something that fully constant folds will tend to be super-fast (techically it should be zero but nothing will say that) and it would not allocate at all.
Which is why we often use `@ballocated` to check for this.
```julia
julia> @btime (()->f((:x,:y)))()
  1.489 ns (0 allocations: 0 bytes)
(:x, :y, :zzz)

julia> @btime (()->g((:x,:y)))()
  107.931 ns (1 allocation: 32 bytes)
(:x, :y, :zzz)
```

Reply:
Thanks for the explanation! I've managed to write the method with `name::Integer` without allocations, but the current implementation of the method where `name::Symbol` is not constant folding for all values of the `name` argument.


It works fine for when `name` is already in `dimnames`:
```
julia> @code_typed (()->NamedDims.expand_dimnames((:x, :y), :x))()
CodeInfo(
1 ─     return (:x, :y)
) => Tuple{Symbol,Symbol}
```

But runs into problems when `name` is not in `dimnames`. It seems like the issue is that it doesn't realise what the return type will be (hence the Union). I don't know how to fix this (it works automagically when `name::Integer`)
```
julia> @code_typed (()->NamedDims.expand_dimnames((:x, :y), :z))()
CodeInfo(
1 ──       goto #12 if not true
2 ┄─ %2  = φ (#1 => :x, #11 => %18)::Symbol
│    %3  = φ (#1 => 2, #11 => %19)::Int64
│    %4  = (%2 === :z)::Bool
└───       goto #4 if not %4
3 ──       goto #13
4 ── %7  = Base.sle_int(1, %3)::Bool
└───       goto #6 if not %7
5 ── %9  = Base.sle_int(%3, 2)::Bool
└───       goto #7
6 ──       nothing::Nothing
7 ┄─ %12 = φ (#5 => %9, #6 => false)::Bool
└───       goto #9 if not %12
8 ── %14 = Base.getfield((:x, :y), %3, false)::Symbol
│    %15 = Base.add_int(%3, 1)::Int64
└───       goto #10
9 ──       goto #10
10 ┄ %18 = φ (#8 => %14)::Symbol
│    %19 = φ (#8 => %15)::Int64
│    %20 = φ (#8 => false, #9 => true)::Bool
│    %21 = Base.not_int(%20)::Bool
└───       goto #12 if not %21
11 ─       goto #2
12 ┄       goto #13
13 ┄ %25 = φ (#3 => true, #12 => false)::Bool
└───       goto #15 if not %25
14 ─       goto #16
15 ─ %28 = Core.tuple(:x, :y, :z)::Tuple{Symbol,Symbol,Symbol}
└───       goto #16
16 ┄ %30 = φ (#14 => (:x, :y), #15 => %28)::Union{Tuple{Symbol,Symbol}, Tuple{Symbol,Symbol,Symbol}}
└───       return %30
) => Union{Tuple{Symbol,Symbol}, Tuple{Symbol,Symbol,Symbol}}
```

Funny enough, if I change the line from `if name in dimnames` to `if false`, then
```
julia> @code_typed (()->NamedDims.expand_dimnames((:x, :y), :z))()
CodeInfo(
1 ─ %1 = Core.tuple(:x, :y, :z)::Tuple{Symbol,Symbol,Symbol}
└──      return %1
) => Tuple{Symbol,Symbol,Symbol}
```

So it looks like the issue (comparing to the other method, where `if name <= length(dimnames)`) has to do with how
the compiler deals with the if statement when there is an iteration involved.

Any ideas on how to fix this?